internal/pkg/iface/iface_linux.go

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2025 k0s authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package iface
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+)
+
+// FirstPublicAddress return the first found non-local IPv4 address that's not part of pod network
+// if any interface does not have any IPv4 address then return the first found non-local IPv6 address
+func FirstPublicAddress() (string, error) {
+	ifs, err := net.Interfaces()
+	if err != nil {
+		return "127.0.0.1", fmt.Errorf("failed to list network interfaces: %w", err)
+	}
+	ipv6addr := ""
+	for _, i := range ifs {
+		if !isPublicInterface(i.Name) {
+			continue
+		}
+
+		link, err := netlink.LinkByName(i.Name)
+		if err != nil {
+			logrus.WithError(err).Warnf("failed to get link by name %s", i.Name)
+			continue
+		}
+
+		addresses, err := netlink.AddrList(link, netlink.FAMILY_ALL)
+		if err != nil {
+			logrus.WithError(err).Warnf("failed to get addresses for interface %s", i.Name)
+			continue
+		}
+
+		for _, a := range addresses {
+			// skip secondary addresses. This is to avoid returning VIPs as the public address
+			// https://github.com/k0sproject/k0s/issues/4664
+			if a.Flags&unix.IFA_F_SECONDARY != 0 {
+				continue
+			}
```

Comment:
Could it happen that there's only secondary addresses? If yes, I'm wondering if we'd rather not skip them entirely, but use them as a fallback if no primary address is found. (Could be done by e.g. using a stable sort on the slice of addresses before iterating over it.)

Reply:
To the best of my knowledge that's not a possible scenario.

The first address added to an interface never has the flag `IFA_F_SECONDARY`. If the primary address is removed, before actually removing it, the kernel either promotes a secondary address or deletes all first. So this cannot happen even if we queried the interfaces in the precise moment while the address is being deleted:

https://github.com/torvalds/linux/blob/09a0fa92e5b45e99cf435b2fbf5ebcf889cf8780/net/ipv4/devinet.c#L396C1-L4546